[setup] Rewrite Ubuntu install_prereqs from the ground up

jwnimmer-tri opened the pull request:
TBD

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RobotLocomotion/drake/22055)
<!-- Reviewable:end -->

jwnimmer-tri reviewed (commented):
> Reviewed 13 of 15 files at r1, 28 of 30 files at r2, 2 of 3 files at r3, 2 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22055)* status: 2 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes

___
*[`doc/_pages/from_source.md` line 42 at r5](https://reviewable.io/reviews/RobotLocomotion/drake/22055#-OC0Vtpl7vLFvH3984HY:-OC0Vtpl7vLFvH3984HZ:b-mb2ndy) ([raw file](https://github.com/RobotLocomotion/drake/blob/94a95dc5228f783a86c7288f5facc870a324a23e/doc/_pages/from_source.md#L42)):*
> ```Markdown
> require a working display server. Most personal computers will have this
> already built in, but some cloud or docker environments may require extra
> setup steps.
> ```

Working

Refresh our advice re: display servers now that we use EGL by default.
___
*[`doc/_pages/installation.md` line 40 at r5](https://reviewable.io/reviews/RobotLocomotion/drake/22055#-OC0VzQQ2gon1ChtyVY6:-OC0VzQQ2gon1ChtyVY7:b-mb2ndy) ([raw file](https://github.com/RobotLocomotion/drake/blob/94a95dc5228f783a86c7288f5facc870a324a23e/doc/_pages/installation.md#L40)):*
> ```Markdown
> require a working display server.  Most personal computers will have this
> already built in, but some cloud or docker environments may require extra
> setup steps.
> ```

Working

Refresh our advice re: display servers now that we use EGL by default.


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> Reviewed 4 of 5 files at r5, 4 of 4 files at r6, 14 of 14 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22055)* status: needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @jwnimmer-tri)

___
*[`doc/_pages/from_source.md` line 42 at r5](https://reviewable.io/reviews/RobotLocomotion/drake/22055#-OC0Vtpl7vLFvH3984HY:-OFTwhJr4y-buXZHq4rz:b9ymizt) ([raw file](https://github.com/RobotLocomotion/drake/blob/94a95dc5228f783a86c7288f5facc870a324a23e/doc/_pages/from_source.md#L42)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Working

Refresh our advice re: display servers now that we use EGL by default.
</blockquote></details>

=> #22365
___
*[`doc/_pages/installation.md` line 40 at r5](https://reviewable.io/reviews/RobotLocomotion/drake/22055#-OC0VzQQ2gon1ChtyVY6:-OFTwjy3348Bgb4Q01KW:b9ymizt) ([raw file](https://github.com/RobotLocomotion/drake/blob/94a95dc5228f783a86c7288f5facc870a324a23e/doc/_pages/installation.md#L40)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Working

Refresh our advice re: display servers now that we use EGL by default.
</blockquote></details>

=> #22365


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> Reviewed 1 of 8 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22055)* status: 1 unresolved discussion, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes

___
*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/22055#-OG1kgRp7R8cOS2N4M25:-OG1kgRp7R8cOS2N4M26:b3fms38) (no related file):*
Working

Document the new solver flags in `bazel.md`.


<!-- Sent from Reviewable.io -->

mwoehlke-kitware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22055)* status: 3 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @jwnimmer-tri)

___
*[`setup/install_prereqs` line 1 at r9](https://reviewable.io/reviews/RobotLocomotion/drake/22055#-OGpR-Km1osUoSXFVIz0:-OGpR-Km1osUoSXFVIz1:bg33601) ([raw file](https://github.com/RobotLocomotion/drake/blob/4cf8eeed1c31f8e83f907fae9fc0e5bff3389cdd/setup/install_prereqs#L1)):*
> ```
> #!/bin/bash
> ```

Huh? This doesn't look like bash script; did you mean `#!/usr/bin/env python3`?
___
*[`setup/ubuntu/packages-jammy-binary.txt` line 1 at r9](https://reviewable.io/reviews/RobotLocomotion/drake/22055#-OGpOufpDJO9BgI_PSb3:-OGpOufpDJO9BgI_PSb4:brjuy0b) ([raw file](https://github.com/RobotLocomotion/drake/blob/4cf8eeed1c31f8e83f907fae9fc0e5bff3389cdd/setup/ubuntu/packages-jammy-binary.txt#L1)):*
> ```help
> build-essential
> ```

You know, probably not in this pass, but eventually, I'd be *very* tempted to turn these into a `pyproject.toml` style single file, e.g.:
```toml
[requirements]
build = [
  "cmake",
  ...
]
test = [
  "kcov; dist == 'jammy'",
   ...
]
```

Or maybe YAML:
```yaml
requirements:
  build:
    - cmake
    - ...
  test:
    - kcov; dist == 'jammy'
    - ...
```
  


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22055)* status: 3 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes

___
*[`setup/install_prereqs` line 1 at r9](https://reviewable.io/reviews/RobotLocomotion/drake/22055#-OGpR-Km1osUoSXFVIz0:-OGpT-4NCMn5iNW_b3y5:b-a6co8g) ([raw file](https://github.com/RobotLocomotion/drake/blob/4cf8eeed1c31f8e83f907fae9fc0e5bff3389cdd/setup/install_prereqs#L1)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

Huh? This doesn't look like bash script; did you mean `#!/usr/bin/env python3`?
</blockquote></details>

See the first comment paragraph immediate below.
___
*[`setup/ubuntu/packages-jammy-binary.txt` line 1 at r9](https://reviewable.io/reviews/RobotLocomotion/drake/22055#-OGpOufpDJO9BgI_PSb3:-OGpT74-BE8-bJ_bMQtL:bbjnycc) ([raw file](https://github.com/RobotLocomotion/drake/blob/4cf8eeed1c31f8e83f907fae9fc0e5bff3389cdd/setup/ubuntu/packages-jammy-binary.txt#L1)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

You know, probably not in this pass, but eventually, I'd be *very* tempted to turn these into a `pyproject.toml` style single file, e.g.:
```toml
[requirements]
build = [
  "cmake",
  ...
]
test = [
  "kcov; dist == 'jammy'",
   ...
]
```

Or maybe YAML:
```yaml
requirements:
  build:
    - cmake
    - ...
  test:
    - kcov; dist == 'jammy'
    - ...
```
</blockquote></details>

It will be simplest if it's something we can parse using only the CPython 3.9 stdlib.  If not, then we first need some other data file to bootstrap whatever parsing library we need.


<!-- Sent from Reviewable.io -->

mwoehlke-kitware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22055)* status: 1 unresolved discussion, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @jwnimmer-tri)

___
*[`setup/install_prereqs` line 1 at r9](https://reviewable.io/reviews/RobotLocomotion/drake/22055#-OGpR-Km1osUoSXFVIz0:-OGqJ8_C4ElTNfNKTg6b:b-406nsw) ([raw file](https://github.com/RobotLocomotion/drake/blob/4cf8eeed1c31f8e83f907fae9fc0e5bff3389cdd/setup/install_prereqs#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

See the first comment paragraph immediate below.
</blockquote></details>

Oh, my, that's... sneaky. Is it really so terrible to have a separate `.py`?
___
*[`setup/ubuntu/packages-jammy-binary.txt` line 1 at r9](https://reviewable.io/reviews/RobotLocomotion/drake/22055#-OGpOufpDJO9BgI_PSb3:-OGqEvCU-A2xN2tSO9oG:bfgiude) ([raw file](https://github.com/RobotLocomotion/drake/blob/4cf8eeed1c31f8e83f907fae9fc0e5bff3389cdd/setup/ubuntu/packages-jammy-binary.txt#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It will be simplest if it's something we can parse using only the CPython 3.9 stdlib.  If not, then we first need some other data file to bootstrap whatever parsing library we need.
</blockquote></details>

Point, or we could declare "our own" format that's pretty dumb, or maybe use JSON. I was thinking TOML went back further than it does, though; I see it's only in 3.11. :slightly_frowning_face: In any case, as I said, not something to muck with *now*.


<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
The work in #24388 to simplify the process will need to land first, as well as subsequent refactoring like renaming and merging the text files.  After that's finished, we can cycle back here to revisit the UX improvements in a Python rewrite.

jwnimmer-tri reviewed (commented):
> @jwnimmer-tri made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/22055)* status: 2 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on mwoehlke-kitware).

___
*[`setup/ubuntu/packages-jammy-binary.txt` line 1 at r9](https://reviewable.io/reviews/robotlocomotion/drake/22055#-OGpOufpDJO9BgI_PSb3:-Oq7Oj_h8llTRcyMtR-V:bw8yfed) ([raw file](https://github.com/robotlocomotion/drake/blob/4cf8eeed1c31f8e83f907fae9fc0e5bff3389cdd/setup/ubuntu/packages-jammy-binary.txt#L1)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

Point, or we could declare "our own" format that's pretty dumb, or maybe use JSON. I was thinking TOML went back further than it does, though; I see it's only in 3.11. :slightly_frowning_face: In any case, as I said, not something to muck with *now*.
</blockquote></details>

Our minimum is 3.12 now, so I might revisit this.


<!-- Sent from Reviewable.io -->

jwnimmer-tri reviewed (commented):
FYI I'm going to work to carve off chunks of this into separate PRs.

> @jwnimmer-tri partially reviewed 14 files and made 2 comments.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/22055#-:-Oqh2aka43GGMQtfzpyP:b-9knh86)* status: 2 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits), missing label for release notes (waiting on mwoehlke-kitware).

___
*[`setup/install_prereqs` line 1 at r9](https://reviewable.io/reviews/robotlocomotion/drake/22055#-OGpR-Km1osUoSXFVIz0:-Oqh2FwU6M5Ok7lha2L1:bc4pcio) ([raw file](https://github.com/robotlocomotion/drake/blob/4cf8eeed1c31f8e83f907fae9fc0e5bff3389cdd/setup/install_prereqs#L1)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

Oh, my, that's... sneaky. Is it really so terrible to have a separate `.py`?
</blockquote></details>

OK yup.  The bash trampoline and python implementation are separate files now.


<!-- Sent from Reviewable.io -->

jwnimmer-tri committed:
PR [setup] Move user environment logic to Python

jwnimmer-tri committed:
[setup] Rewrite Ubuntu install_prereqs from the ground up

Instead, users and developers should run setup/install_prereqs (not as
root).

Usability changes:
- Users never run any of our commands under sudo; we'll request sudo if
  and only if we actually need superuser access.
- Fine-tuned log output, with option for --verbose (e.g., in CI).
- Skip unnecessary steps; incremental runs are typically sub-second.

Maintainability changes:
- One single-file program to maintain, instead of an assortment of bash
  scripts and source and exec each other in weird ways.
- Always use files, not heredocs. (The pkg-config etc in a heredoc were
  documented as unwanted when using Drake's Debian packages, but that
  was wrong; we always want those deps even in Debian packages.) All of
  the new entries in our txt files are ported over from heredocs, as is
  the bazelisk metadata.
- Flavors are now a linear stack, not pick-and-choose hodgepodge. This
  substantially reduces our testing / verification burden. Filenames of
  txt files now exactly match the flavor names.
- Smoother control flow which should pave the way to supporting more
  kinds of platforms beyond Ubuntu. Adding more and better kinds of
  error handling or reporting will be easier now (vs bash).
